snewpdag/plugins/gen/TrueTimeDelay.py

```diff
@@ -0,0 +1,121 @@
+"""
+TrueTimeDelay: Calculate the true time delay of the neutrino arrival between two detectors
+by randomly generate a neutrino directional vector and a arrival time 
+
+Constructor Arguments:
+    detector_list: list of strings, ["first_detector", "second_detector", ...]
+                the list of detectors that we want to generate true time delay
+                options: "HK", "IC", "JUNO", "KM3", "SK"
+    detector_location: csv file ('detector_location.csv')
+
+Output added to data: dictionary with key ("det1", "det2") and the corresponding true time delay
+
+"""
+
+import csv
+import time
+import numpy as np
+import random as rm
+import itertools
+from datetime import datetime
+from snewpdag.dag import Node
+
+class TrueTimeDelay(Node):
+
+    #Define detector location
+    def __init__(self, detector_list, detector_location, **kwargs):
+        self.detector_info = {}
+        with open(detector_location, 'r') as f:
+            detectors = csv.reader(f)
+            for detector in detectors:
+                name = detector[0]
+                if name in detector_list: 
+                    lon = np.radians(float(detector[1]))
+                    lat = np.radians(float(detector[2]))
+                    height = float(detector[3])
+                    self.detector_info[name] = [lon, lat, height]
+        super().__init__(**kwargs)
+
+
+    #Randomly generate the direction vector for incoming neutrino flux, using right-ascention(alpha) and declination(delta)
+    def generate_n(self):
+        alpha_deg = rm.uniform(-180,180)
+        alpha = np.radians(alpha_deg)
+
+        delta_distribution = rm.uniform(-1,1)
+        delta = np.arccos(delta_distribution)
+        
+        nx = -np.cos(alpha)*np.cos(delta)
```

Comment:
whoops - it's the polar angle (angle from z axis) which is uniform in its cosine.  I forgot that declination is measured from the equatorial plane.  So one should still throw a random number for polar angle, convert to nx,ny,nz (don't need to arccos and then cos, just use the random number directly where possible).  Convert to declination if necessary.